src/plugins/vis_builder/public/application/components/data_tab/field_selector.tsx

```diff
@@ -65,6 +93,63 @@ export const FieldSelector = () => {
     [filteredFields]
   );
 
+  useEffect(() => {
+    async function getData() {
+      if (indexPattern && searchContext) {
+        const newSearchSource = await searchSource.create();
+        const timeRangeFilter = timefilter.createFilter(indexPattern);
+
+        newSearchSource
+          .setField('index', indexPattern)
+          .setField('size', config.get('discover:sampleSize') ?? 500)
+          .setField('sort', [{ [indexPattern.timeFieldName || '_score']: 'desc' as SortDirection }])
+          .setField('filter', [
+            ...(searchContext.filters ?? []),
+            ...(timeRangeFilter ? [timeRangeFilter] : []),
+          ]);
+
+        if (searchContext.query) {
+          const contextQuery =
+            searchContext.query instanceof Array ? searchContext.query[0] : searchContext.query;
+
+          newSearchSource.setField('query', contextQuery);
+        }
+
+        const searchResponse = await newSearchSource.fetch();
+
+        setHits(searchResponse.hits.hits);
+      }
+    }
+
+    getData();
+  }, [config, searchContext, searchSource, indexPattern, timefilter]);
+
+  useLayoutEffect(() => {
+    const subscription = state$.subscribe(({ state }) => {
+      setSearchContext({
+        query: state.query,
+        filters: state.filters,
+      });
+    });
+
+    return () => {
+      subscription.unsubscribe();
+    };
+  }, [state$]);
```

Comment:
Yeah - that's what we talked about, and where I got it from 😁. I can refactor it into a single hook.

Reply:
No need to refactor until after we complete https://github.com/opensearch-project/OpenSearch-Dashboards/issues/2780, as it may no longer be necessary.